### PR TITLE
net-libs/webkit-gtk: Version bump to 3.38.0

### PR DESCRIPTION
**Pull request by listout**
The slotting is to handle libsoup3 and libsoup2. Some packages from GNOME 43 are requiring webkit2gtk-4.1 (gtk3 with libsoup3), for example GNOME boxes.

Signed-off-by: brahmajit das <listout@protonmail.com>

### CODE REVIEW

**Comment by listout** [comment]
@felixapitzsch I was not aware of that, unfortunately. Thanks for pointing to the cmake gstreamer file, let me investigate a bit

**Comment by mattst88** [comment]
Version should be 2.38.0, not 3.38.0. :)

**Comment by mattst88** [comment]
I see this with `-r0` (will check with -r410):

````
* One or more CMake variables were not used by the project:
*   ENABLE_GTKDOC
*   USE_LIBNOTIFY
````

**Comment by listout** [comment]
> I see this with `-r0` (will check with -r410):
> 
> ```
> * One or more CMake variables were not used by the project:
> *   ENABLE_GTKDOC
> *   USE_LIBNOTIFY
> ```

Yes, I've noticed this too.

**Comment by mattst88** [comment]
Yep, same with `-r410`.

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+inherit check-reqs flag-o-matic gnome2 optfeature python-any-r1 ruby-single toolchain-funcs cmake
+
+MY_P="webkitgtk-${PV}"
+DESCRIPTION="Open source web browser engine"
+HOMEPAGE="https://www.webkitgtk.org"
+SRC_URI="https://www.webkitgtk.org/releases/${MY_P}.tar.xz"
+
+LICENSE="LGPL-2+ BSD"
+SLOT="4.1" # soname version of libwebkit2gtk-4.0
```
Comment (mattst88) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:17]: I suspect we'll want a subslot here to be safe. (Also, fix the mention of `libwebkit2gtk-4.0` to `libwebkit2gtk-4.1`).

I see that the shared objects SONAMEs changed like this:

````diff
-./usr/lib64/libjavascriptcoregtk-4.0.so.18.21.3
-./usr/lib64/libwebkit2gtk-4.0.so.37.57.3
+./usr/lib64/libjavascriptcoregtk-4.1.so.0.2.3
+./usr/lib64/libwebkit2gtk-4.1.so.0.2.3
````

So I think it should be `SLOT="4.1/0".

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r0.ebuild [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+		-DENABLE_GTKDOC=$(usex gtk-doc gi-docgen)
+		-DENABLE_INTROSPECTION=$(usex introspection)
+		-DENABLE_JOURNALD_LOG=$(usex systemd)
+		-DENABLE_QUARTZ_TARGET=$(usex aqua)
+		-DENABLE_WAYLAND_TARGET=$(usex wayland)
+		-DENABLE_X11_TARGET=$(usex X)
+		-DUSE_ANGLE_WEBGL=OFF
+		-DUSE_AVIF=$(usex avif)
+		-DUSE_GTK4=OFF
+		-DENABLE_WEBDRIVER=OFF # Disable WebDriver for webkit2gtk-4.0 and use the webkit2gtk-4.1
```
Comment (mattst88) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r0.ebuild:233]: Let's explicitly enable this in the `-r410` ebuild.

**Review by mattst88 — CHANGES_REQUESTED** [review]

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild (hunk 1 above)
Comment (leio) [reply net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:17]: I don't expect any ABI breaks here anymore, it was a one-time exceptional thing back in CVS times. It remained 4/37 for the whole time in git.
I don't mind either way for 4.1, but I am 99,99% sure it won't change ABI for the gtk4 version that's planned for SLOT=5, as we'll get a 5.1 or 6 or something in March 2023 instead.

**Comment by leio** [comment]
I think we should just unconditionally install the pregenerated gtk-docs in SLOT=4.1, not install any API docs in SLOT=4.0 and call that a day. The API is the same, so whoever actually wants to read it offline from this will be able to dev on top of libsoup:2 with the libsoup:3 versions API docs in devhelp.
SLOT=5 will have a different API and we might want to expose a IUSE=gtk-doc there when that API isn't pregenerated in the tarball, but this isn't a topic for the PR at hand here.

That would mean passing ENABLE_GTKDOC in both slots and doing the doins to /usr/share/gtk-doc/html in SLOT=4.1 only

**Comment by mattst88** [comment]
I see this in the `pkgdiff` output between 2.36.8 and 2.38:

````
 # Public options specific to the GTK port. Do not add any options here unless
 # there is a strong reason we should support changing the value of the option,
-# and the option is not relevant to any other WebKit ports.
+# and the option is not relevant to other WebKit ports.
+WEBKIT_OPTION_DEFINE(ENABLE_DOCUMENTATION "Whether to generate documentation." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(ENABLE_GLES2 "Whether to enable OpenGL ES 2.0." PUBLIC ${ENABLE_GLES2_DEFAULT})
-WEBKIT_OPTION_DEFINE(ENABLE_GTKDOC "Whether or not to use generate gtkdoc." PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(ENABLE_INTROSPECTION "Whether to enable GObject introspection." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(ENABLE_JOURNALD_LOG "Whether to enable journald logging" PUBLIC ON)
 WEBKIT_OPTION_DEFINE(ENABLE_QUARTZ_TARGET "Whether to enable support for the Quartz windowing target." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(ENABLE_WAYLAND_TARGET "Whether to enable support for the Wayland windowing target." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(ENABLE_X11_TARGET "Whether to enable support for the X11 windowing target." PUBLIC ON)
-WEBKIT_OPTION_DEFINE(USE_ANGLE_WEBGL "Whether to use ANGLE as WebGL backend." PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(USE_AVIF "Whether to enable support for AVIF images." PUBLIC ${ENABLE_EXPERIMENTAL_FEATURES})
 WEBKIT_OPTION_DEFINE(USE_GTK4 "Whether to enable usage of GTK4 instead of GTK3." PUBLIC OFF)
 WEBKIT_OPTION_DEFINE(USE_JPEGXL "Whether to enable support for JPEG-XL images." PUBLIC ${ENABLE_EXPERIMENTAL_FEATURES})
 WEBKIT_OPTION_DEFINE(USE_LCMS "Whether to enable support for image color management using libcms2." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBHYPHEN "Whether to enable the default automatic hyphenation implementation." PUBLIC ON)
-WEBKIT_OPTION_DEFINE(USE_LIBNOTIFY "Whether to enable the default web notification implementation." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_LIBSECRET "Whether to enable the persistent credential storage using libsecret." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_OPENGL_OR_ES "Whether to use OpenGL or ES." PUBLIC ON)
 WEBKIT_OPTION_DEFINE(USE_OPENJPEG "Whether to enable support for JPEG2000 images." PUBLIC ON)
````

So we need to stop using the `ENABLE_GTKDOC`, `USE_ANGLE_WEBGL`, and `USE_LIBNOTIFY` options, and use the `ENABLE_DOCUMENTATION` option (which you're doing).

In general, we want to minimize the differences between the `-r0` and `-r410` ebuilds to /just/ what we have to have. E.g., if we want documentation to be provided by one slot and not the other, then one should use `ENABLE_DOCUMENTATION=ON` and the other should use `ENABLE_DOCUMENTATION=OFF`. It's not good to just set it in one ebuild and not the other.

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild (hunk 1 above)
Comment (leio) [reply net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:17]: As per IRC discussion, just go with `SLOT="4.1/0"` then

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r0.ebuild [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+MY_P="webkitgtk-${PV}"
+DESCRIPTION="Open source web browser engine"
+HOMEPAGE="https://www.webkitgtk.org"
+SRC_URI="https://www.webkitgtk.org/releases/${MY_P}.tar.xz"
+
+LICENSE="LGPL-2+ BSD"
+SLOT="4/37" # soname version of libwebkit2gtk-4.0
+KEYWORDS="~amd64 ~arm ~arm64 ~ppc ~ppc64 ~riscv ~sparc ~x86"
+
+IUSE="aqua avif +egl examples gamepad gles2-only gnome-keyring +gstreamer gtk-doc +introspection +jpeg2k +jumbo-build lcms seccomp spell systemd test wayland X"
```
Comment (leio) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r0.ebuild:20]: Nothing is done with IUSE=gtk-doc in this revision anymore, so it shouldn't be there

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r0.ebuild [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+
+	>=dev-lang/perl-5.10
+	virtual/perl-Data-Dumper
+	virtual/perl-Carp
+	virtual/perl-JSON-PP
+
+	gtk-doc? (
+		>=dev-util/gi-docgen-2021.1
+		>=dev-util/gtk-doc-1.32
+	)
```
Comment (leio) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r0.ebuild:125]: ENABLE_DOCUMENTATION=OFF is passed, so this shouldn't be here then. Also it's never both gi-docgen and gtk-doc. And if it is using gtk-doc, then it almost always also needs one or more slots of docbook-xml-dtd.

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+		-DENABLE_SPELLCHECK=$(usex spell)
+		-DENABLE_UNIFIED_BUILDS=$(usex jumbo-build)
+		-DENABLE_VIDEO=$(usex gstreamer)
+		-DENABLE_WEBGL=ON
+		# Supported only under ANGLE
+		-DENABLE_WEBGL2=OFF
+		-DENABLE_WEB_AUDIO=$(usex gstreamer)
+		# Source/cmake/OptionsGTK.cmake
+		-DENABLE_GLES2=$(usex gles2-only)
+		-DENABLE_DOCUMENTATION=ON
```
Comment (leio) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:221]: And now in -r410 you are enabling documentation, but not having the necessary build dependency for it (gi-docgen).
BUT, as I've told, it shouldn't be building documentation in any version, and just install the pregenerated ones in SLOT=4.1 version (-r410) in src_install, similar to what gui-libs/gtk does.

IF you build it, then there's no reason to hide the option to build it for SLOT=4 (-r0) as well, but I strongly disagree with building anything when there's pregenerated ones available in the release tarball.

**Review by leio — CHANGES_REQUESTED** [review]
Additionally your ebuild name is wrong - `-r0` is implicit and must NOT be in the filename; so webkit-gtk-2.38.0-r0.ebuild needs to be renamed webkit-gtk-2.38.0.ebuild

Also this isn't a full review, just what I noticed based on gtk-doc and surrounding stuff due to discussions on IRC.

**Comment by thesamesam** [comment]
See also https://bugs.gentoo.org/872581 (and add a Closes: tag to your commit). Thanks!

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+
+# Tests fail to link for inexplicable reasons
+# https://bugs.webkit.org/show_bug.cgi?id=148210
+RESTRICT="test"
+
+# Dependencies found at Source/cmake/OptionsGTK.cmake
+# Missing WebRTC support, but ENABLE_MEDIA_STREAM/ENABLE_WEB_RTC is experimental upstream (PRIVATE OFF) and shouldn't be used yet in 2.30
+# >=gst-plugins-opus-1.14.4-r1 for opusparse (required by MSE)
+# TODO: gst-plugins-base[X] is only needed when build configuration ends up with GLX set, but that's a bit automagic too to fix
+# Softblocking webkit-2.38:4 as we going to use webkit-2.38:4.1's WebKitDriver binary
```
Comment (mattst88) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:38]: Should be `s/webkit-2.38/webkit-gtk-2.38/g` here

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+	seccomp? (
+		>=sys-apps/bubblewrap-0.3.1
+		sys-libs/libseccomp
+		sys-apps/xdg-dbus-proxy
+	)
+
+	systemd? ( sys-apps/systemd:= )
+	gamepad? ( >=dev-libs/libmanette-0.2.4 )
+	!<net-libs/webkit-2.38:4
```
Comment (mattst88) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:103]: Should be `s/webkit-2.38/webkit-gtk-2.38/g` here

File: net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+		-DUSE_LIBHYPHEN=ON
+		-DUSE_LIBSECRET=$(usex gnome-keyring)
+		-DUSE_OPENGL_OR_ES=ON
+		-DUSE_OPENJPEG=$(usex jpeg2k)
+		-DUSE_SOUP2=OFF
+		-DUSE_WOFF2=ON
+		-DUSE_WPE_RENDERER=$(usex wayland) # WPE renderer is used to implement accelerated compositing under wayland
+		# Disabling PDF.js in webkit-gtk-2.38.x as the eventually-solution to treating PDF files like any other download in epiphany,
+		# please refer bug 872581 and https://gitlab.gnome.org/GNOME/epiphany/-/issues/1750
+		-DENABLE_PDFJS=ON
```
Comment (mattst88) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0-r410.ebuild:241]: It's been requested that we make this optional. I'm okay with that.

File: net-libs/webkit-gtk/webkit-gtk-2.38.0.ebuild [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+		-DUSE_LIBHYPHEN=ON
+		-DUSE_LIBSECRET=$(usex gnome-keyring)
+		-DUSE_OPENGL_OR_ES=ON
+		-DUSE_OPENJPEG=$(usex jpeg2k)
+		-DUSE_SOUP2=ON
+		-DUSE_WOFF2=ON
+		-DUSE_WPE_RENDERER=$(usex wayland) # WPE renderer is used to implement accelerated compositing under wayland
+		# Disabling PDF.js in webkit-gtk-2.38.x as the eventually-solution to treating PDF files like any other download in epiphany,
+		# please refer bug 872581 and https://gitlab.gnome.org/GNOME/epiphany/-/issues/1750
+		-DENABLE_PDFJS=ON
```
Comment (mattst88) [inline net-libs/webkit-gtk/webkit-gtk-2.38.0.ebuild:238]: Same here.

**Comment by orlitzky** [comment]
Thank you!

**Pull request closed without merging**